OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,286 @@
+
+/** \file
+This file defines the  DataTable_ class, which is used by OpenSim to provide an 
+in-memory container for data access and manipulation.                         */
+
+#ifndef OPENSIM_DATATABLE_H
+#define OPENSIM_DATATABLE_H
+
+// Non-standard headers.
+#include "SimTKcommon.h"
+#include "OpenSim/Common/Exception.h"
+#include "OpenSim/Common/ValueArrayDictionary.h"
+
+namespace OpenSim {
+
+/** AbstractDataTable is the base-class of all DataTable_(templated) allowing 
+storage of DataTable_ templated on different types to be stored in a container 
+like std::vector. AbstractDataTable_ offers:
+- Interface to access column metadata of the DataTable_.
+- A heterogeneous container to store metadata associated with the DataTable_ in
+  the form of key-value pairs where key is of type std::string and value can be
+  of any type.
+
+This class is abstract and cannot be used directly. Create instances of 
+DataTable_ instead. See DataTable_ for details on ussage.                     */
+class AbstractDataTable {
+protected:
+    class ValueDictionary {
+    public:
+        using AbstractValue = SimTK::AbstractValue;
+
+        const AbstractValue& getValueForKey(const std::string& key) const {
+            return _dictionary.getValueArrayForKey(key)[0];
+        }
+
+        template<typename ValueType>
+        void setValueForKey(const std::string& key,
+                            const ValueType& value) {
+            ValueArray<ValueType> value_array{};
+            value_array.upd().push_back(SimTK::Value<ValueType>{value});
+            _dictionary.setValueArrayForKey(key, value_array);
+        }
+
+        void removeValueForKey(const std::string& key) {
+            _dictionary.removeValueArrayForKey(key);
+        }
+
+        std::vector<std::string> getKeys() const {
+            return _dictionary.getKeys();
+        }
+
+    private:
+        ValueArrayDictionary _dictionary;
+    }; // ValueDictionary
+
+public:
+    using TableMetaData       = ValueDictionary;
+    using DependentsMetaData  = ValueArrayDictionary;
+    using IndependentMetaData = ValueDictionary;
+
+    AbstractDataTable()                                      = default;
+    AbstractDataTable(const AbstractDataTable&)              = default;
+    AbstractDataTable(AbstractDataTable&&)                   = default;
+    AbstractDataTable& operator=(const AbstractDataTable&)   = default;
+    AbstractDataTable& operator=(AbstractDataTable&&)        = default;
+    virtual std::unique_ptr<AbstractDataTable> clone() const = 0;
+    virtual ~AbstractDataTable()                             = default;
+
+    /** Get metadata associated with the table.                               */
+    const TableMetaData& getTableMetaData() const {
+        return _tableMetaData;
+    }
+
+    /** Update metadata associated with the table.                            */
+    TableMetaData& updTableMetaData() {
```

Comment:
This returns a protected type from a public method. Does the public `using` promote `ValueDictionary` to public?